analytic_engine/src/instance/wal_replayer.rs

```diff
@@ -0,0 +1,601 @@
+
+        // TODO: Replay logs of different tables in parallel.
+        for table_batch in table_batches {
+            // Some tables may have failed in previous replay, ignore them.
+            if faileds.contains_key(&table_batch.table_id) {
+                continue;
+            }
+
+            // Replay all log entries of current table.
+            // Some tables may have been moved to other shards or dropped, ignore such logs.
+            if let Some(ctx) = serial_exec_ctxs.get_mut(&table_batch.table_id) {
+                let result = replay_table_log_entries(
+                    &context.flusher,
+                    context.max_retry_flush_limit,
+                    &mut ctx.serial_exec,
+                    &ctx.table_data,
+                    log_batch.range(table_batch.start_log_idx..table_batch.end_log_idx),
+                )
+                .await;
+
+                // If occur error, mark this table as failed and store the cause.
+                if let Err(e) = result {
+                    faileds.insert(table_batch.table_id, e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn split_log_batch_by_table<P>(
+        log_batch: &VecDeque<LogEntry<P>>,
+        table_batches: &mut Vec<TableBatch>,
+    ) {
+        table_batches.clear();
+
+        if log_batch.is_empty() {
+            return;
+        }
+
+        // Split log batch by table id, for example:
+        // input batch:
+        //  |1|1|2|2|2|3|3|3|3|
+        //
+        // output batches:
+        //  |1|1|, |2|2|2|, |3|3|3|3|
+        let mut start_log_idx = 0usize;
+        let mut curr_log_idx = 0usize;
+        let mut start_table_id = log_batch.get(start_log_idx).unwrap().table_id;
+        loop {
+            let time_to_break = curr_log_idx == log_batch.len();
+            let found_end_idx = if time_to_break {
+                true
+            } else {
+                let current_table_id = log_batch.get(curr_log_idx).unwrap().table_id;
+                current_table_id != start_table_id
+            };
+
+            if found_end_idx {
+                table_batches.push(TableBatch {
+                    table_id: TableId::new(start_table_id),
+                    start_log_idx,
+                    end_log_idx: curr_log_idx,
+                });
+
+                // Step to next start idx.
+                start_log_idx = curr_log_idx;
+                start_table_id = if time_to_break {
+                    // The final round, just set it to max as an invalid flag.
+                    u64::MAX
+                } else {
+                    log_batch.get(start_log_idx).unwrap().table_id
+                };
+            }
+
+            if time_to_break {
+                break;
+            }
+            curr_log_idx += 1;
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct TableBatch {
+    table_id: TableId,
+    start_log_idx: usize,
```

Comment:
Use [range](http://doc.rust-lang.org/1.69.0/std/ops/struct.Range.html) here, it will tell user which idx in inclusive or not.